diff -up nfs-utils-1.3.0/utils/exportfs/exports.man.orig nfs-utils-1.3.0/utils/exportfs/exports.man
--- nfs-utils-1.3.0/utils/exportfs/exports.man.orig	2017-11-03 07:45:47.000000000 -0400
+++ nfs-utils-1.3.0/utils/exportfs/exports.man	2017-11-03 09:25:47.000000000 -0400
@@ -410,7 +410,7 @@ of the filesystem must be handled elsewh
 
 .TP
 .IR pnfs
-This option allows enables the use of pNFS extension if protocol level
+This option enables the use of the pNFS extension if the protocol level
 is NFSv4.1 or higher, and the filesystem supports pNFS exports.  With
 pNFS clients can bypass the server and perform I/O directly to storage
 devices. The default can be explicitly requested with the
diff -up nfs-utils-1.3.0/utils/mount/nfs.man.orig nfs-utils-1.3.0/utils/mount/nfs.man
--- nfs-utils-1.3.0/utils/mount/nfs.man.orig	2017-11-03 07:45:47.000000000 -0400
+++ nfs-utils-1.3.0/utils/mount/nfs.man	2017-11-03 09:23:20.000000000 -0400
@@ -11,11 +11,8 @@ NFS is an Internet Standard protocol
 created by Sun Microsystems in 1984. NFS was developed
 to allow file sharing between systems residing
 on a local area network.
-The Linux NFS client supports three versions
-of the NFS protocol:
-NFS version 2 [RFC1094],
-NFS version 3 [RFC1813],
-and NFS version 4 [RFC3530].
+Depending on kernel configuration, the Linux NFS client may
+support NFS versions 2, 3, 4.0, 4.1, or 4.2.
 .P
 The
 .BR mount (8)
@@ -88,9 +85,8 @@ These options are valid to use with any
 The NFS protocol version number used to contact the server's NFS service.
 If the server does not support the requested version, the mount request 
 fails.
-If this option is not specified, the client negotiates a suitable version 
-with
-the server, trying version 4 first, version 3 second, and version 2 last.
+If this option is not specified, the client tries version 4.1 first,
+then negotiates down until it finds a version supported by the server.
 .TP 1.5i
 .BI vers= n
 This option is an alternative to the
@@ -752,7 +748,7 @@ NOTE: When used together, the 'local_loc
 by 'nolock'/'lock' mount option.
 .SS "Options for NFS version 4 only"
 Use these options, along with the options in the first subsection above,
-for NFS version 4 and newer.
+for NFS version 4.0 and newer.
 .TP 1.5i
 .BI proto= netid
 The
@@ -825,7 +821,7 @@ the behavior of this option in more deta
 Specifies a single IPv4 address (in dotted-quad form),
 or a non-link-local IPv6 address,
 that the NFS client advertises to allow servers
-to perform NFS version 4 callback requests against
+to perform NFS version 4.0 callback requests against
 files on this mount point. If  the  server is unable to
 establish callback connections to clients, performance
 may degrade, or accesses to files may temporarily hang.
@@ -838,6 +834,11 @@ In the presence of multiple client netwo
 special routing policies,
 or atypical network topologies,
 the exact address to use for callbacks may be nontrivial to determine.
+.IP
+NFS protocol versions 4.1 and 4.2 use the client-established
+TCP connection for callback requests, so do not require the server to
+connect to the client.  This option is therefore only affect NFS version
+4.0 mounts.
 .TP 1.5i
 .BR migration " / " nomigration
 Selects whether the client uses an identification string that is compatible
@@ -1726,7 +1727,7 @@ file system table
 .TP 1.5i
 .I /etc/nfsmount.conf
 Configuration file for NFS mounts
-.SH BUGS
+.SH NOTES
 Before 2.4.7, the Linux NFS client did not support NFS over TCP.
 .P
 Before 2.4.20, the Linux NFS client used a heuristic
@@ -1745,9 +1746,9 @@ when the
 .BR rsize " and " wsize
 settings were smaller than the system's page size.
 .P
-The Linux NFS client does not yet support
-certain optional features of the NFS version 4 protocol,
-such as security negotiation, server referrals, and named attributes.
+The Linux client's support for protocol versions depend on whether the
+kernel was built with options CONFIG_NFS_V2, CONFIG_NFS_V3,
+CONFIG_NFS_V4, CONFIG_NFS_V4_1, and CONFIG_NFS_V4_2.
 .SH "SEE ALSO"
 .BR fstab (5),
 .BR mount (8),
@@ -1780,4 +1781,8 @@ RFC 1833 for the RPC bind specification.
 .br
 RFC 2203 for the RPCSEC GSS API protocol specification.
 .br
-RFC 3530 for the NFS version 4 specification.
+RFC 7530 for the NFS version 4.0 specification.
+.br
+RFC 5661 for the NFS version 4.1 specification.
+.br
+RFC 7862 for the NFS version 4.2 specification.
