diff -up nfs-utils-1.3.0/systemd/nfs-server.service.orig nfs-utils-1.3.0/systemd/nfs-server.service
--- nfs-utils-1.3.0/systemd/nfs-server.service.orig	2017-12-12 10:16:47.164190963 -0500
+++ nfs-utils-1.3.0/systemd/nfs-server.service	2017-12-12 10:54:16.671323896 -0500
@@ -25,8 +25,8 @@ EnvironmentFile=-/run/sysconfig/nfs-util
 Type=oneshot
 RemainAfterExit=yes
 ExecStartPre=-/usr/sbin/exportfs -r
-ExecStartPre=-/bin/sh -c '/bin/kill -HUP `cat /run/gssproxy.pid`'
 ExecStart=/usr/sbin/rpc.nfsd $RPCNFSDARGS
+ExecStartPost=-/bin/sh -c 'if systemctl -q is-active gssproxy; then systemctl restart gssproxy ; fi'
 ExecStop=/usr/sbin/rpc.nfsd 0
 ExecStopPost=/usr/sbin/exportfs -au
 ExecStopPost=/usr/sbin/exportfs -f
