diff -up nfs-utils-1.2.3/utils/mount/nfsmount.conf.orig nfs-utils-1.2.3/utils/mount/nfsmount.conf
--- nfs-utils-1.2.3/utils/mount/nfsmount.conf.orig	2014-04-29 14:17:07.588350000 -0400
+++ nfs-utils-1.2.3/utils/mount/nfsmount.conf	2014-04-29 14:33:52.318087000 -0400
@@ -133,3 +133,12 @@
 # RPCGSS security flavors 
 # [none, sys, krb5, krb5i, krb5p ]
 # Sec=sys
+#
+# Allow Signals to interrupt file operations
+# Intr=True
+#
+# Specifies  how the kernel manages its cache of directory
+# Lookupcache=all|none|pos|positive
+#
+# Turn of the caching of that access time
+# noatime=True
