diff -up nfs-utils-1.3.0/utils/statd/start-statd.orig nfs-utils-1.3.0/utils/statd/start-statd
--- nfs-utils-1.3.0/utils/statd/start-statd.orig	2014-09-30 09:17:31.000000000 -0400
+++ nfs-utils-1.3.0/utils/statd/start-statd	2014-09-30 09:15:01.000000000 -0400
@@ -5,8 +5,9 @@
 # It should run statd with whatever flags are apropriate for this
 # site.
 PATH="/sbin:/usr/sbin:/bin:/usr/bin"
-if systemctl start statd.service
+if systemctl start rpc-statd.service
 then :
 else
     exec rpc.statd --no-notify
 fi
+
