diff -up nfs-utils-1.2.3/utils/exportfs/exportfs.c.orig nfs-utils-1.2.3/utils/exportfs/exportfs.c
--- nfs-utils-1.2.3/utils/exportfs/exportfs.c.orig	2015-12-19 08:16:25.536590551 -0500
+++ nfs-utils-1.2.3/utils/exportfs/exportfs.c	2015-12-19 08:21:58.328265316 -0500
@@ -488,21 +488,25 @@ static int can_test(void)
 static int test_export(char *path, int with_fsid)
 {
 	char buf[1024];
+	char *bp = buf;
+	int len = sizeof(buf);
 	int fd, n;
 
-	n = snprintf(buf, 1024, "-test-client- %s 3 %d -1 -1 0\n",
-		path,
-		with_fsid ? NFSEXP_FSID : 0);
-	if (n >= 1024)
-		return 0;
+	n = snprintf(buf, len, "-test-client- ");
+	bp += n;
+	len -= n;
+	qword_add(&bp, &len, path);
+	snprintf(bp, len, " 3 %d 65534 65534 0\n", with_fsid ? NFSEXP_FSID : 0);
 
 	fd = open("/proc/net/rpc/nfsd.export/channel", O_WRONLY);
 	if (fd < 0)
 		return 0;
+
 	n = write(fd, buf, strlen(buf));
 	close(fd);
 	if (n < 0)
 		return 0;
+
 	return 1;
 }
 
