diff -up nfs-utils-1.2.3/utils/mount/nfsumount.c.orig nfs-utils-1.2.3/utils/mount/nfsumount.c
--- nfs-utils-1.2.3/utils/mount/nfsumount.c.orig	2012-04-25 12:37:54.043289000 -0400
+++ nfs-utils-1.2.3/utils/mount/nfsumount.c	2012-04-25 14:12:02.979473000 -0400
@@ -139,53 +139,6 @@ static int del_mtab(const char *spec, co
 }
 
 /*
- * To ensure the path is found during unmounts, strip
- * off the multiple '/' or add a '/' if one does not exist.
- */
-static inline char *
-normalize_path(char *spec) 
-{
-	char *colen, *ptr, *str, *dev = NULL;
-
-	if ((colen = strchr(spec, ':')) == NULL)
-		return NULL;
-
-	if (*(colen + 1) != '/') {
-		dev = str = malloc(strlen(spec) + 2);
-		if (dev == NULL) 
-			return NULL;
-
-		ptr = spec;
-		while (ptr <= colen)
-			*(str++) = *(ptr++);
-		*str++='/';
-		while (*ptr)
-			*(str++) = *(ptr++);
-		*str='\0';
-
-		return dev;
-	}
-
-	if ((str = strstr(spec, "//")) != NULL) {
-		if (dev == NULL) {
-			dev = strdup(spec);
-			if (dev == NULL)
-				return NULL;
-		}
-		while ((str = strstr(dev, "//")) != NULL) {
-			ptr = str;
-			while (*ptr && *ptr == '/')
-				ptr++;
-			str++;
-			while (*ptr)
-				*(str++) = *(ptr++);
-			*str = '\0';
-		}
-	} 
-	return dev;
-}
-
-/*
  * Discover mount server's hostname/address by examining mount options
  *
  * Returns a pointer to a string that the caller must free, on
@@ -315,30 +268,6 @@ static int nfs_umount_is_vers4(const str
 		goto not_found;
 
 	do {
-		int nlen = strlen(pmc->m.mnt_fsname);
-		/*
-		 * It's possible the mount location string in /proc/mounts
-		 * ends with a '/'. In this case, if the entry came from
-		 * /etc/mtab, it won't have the trailing '/' so deal with
-		 * it.
-		 */
-		while (pmc->m.mnt_fsname[nlen - 1] == '/')
-			nlen--;
-		/*
-		 * When the mtab and /proc/mounts are not the same
-		 * file, normalize the path in the mtab if needed.
-		 */
-		if (mtab_is_writable())
-			normpath = normalize_path(mc->m.mnt_fsname);
-
-		if (strncmp(pmc->m.mnt_fsname, mc->m.mnt_fsname, nlen) != 0) {
-			/* Is there a normalized path, if so compare that one too */
-			if (normpath == NULL)
-				continue;
-			if (strncmp(pmc->m.mnt_fsname, normpath, nlen) != 0)
-				continue;
-		}
-
 		if (strcmp(pmc->m.mnt_type, "nfs4") == 0)
 			goto out_nfs4;
 
