diff -up nfs-utils-1.2.3/utils/exportfs/exportfs.c.orig nfs-utils-1.2.3/utils/exportfs/exportfs.c
--- nfs-utils-1.2.3/utils/exportfs/exportfs.c.orig	2014-05-02 12:58:43.000000000 -0400
+++ nfs-utils-1.2.3/utils/exportfs/exportfs.c	2014-05-02 13:08:31.000000000 -0400
@@ -394,9 +394,12 @@ static int test_export(char *path, int w
 	char buf[1024];
 	int fd, n;
 
-	sprintf(buf, "-test-client- %s 3 %d -1 -1 0\n",
+	n = snprintf(buf, 1024, "-test-client- %s 3 %d -1 -1 0\n",
 		path,
 		with_fsid ? NFSEXP_FSID : 0);
+	if (n >= 1024)
+		return 0;
+
 	fd = open("/proc/net/rpc/nfsd.export/channel", O_WRONLY);
 	if (fd < 0)
 		return 0;
