diff -up nfs-utils-1.2.3/support/nfs/conffile.c.orig nfs-utils-1.2.3/support/nfs/conffile.c
--- nfs-utils-1.2.3/support/nfs/conffile.c.orig	2012-03-06 10:39:47.938798000 -0500
+++ nfs-utils-1.2.3/support/nfs/conffile.c	2012-03-06 11:02:35.077261000 -0500
@@ -211,7 +211,7 @@ static void
 conf_parse_line(int trans, char *line, size_t sz)
 {
 	char *val, *ptr;
-	size_t i;
+	size_t i, valsize;
 	size_t j;
 	static char *section = 0;
 	static char *arg = 0;
@@ -296,23 +296,16 @@ conf_parse_line(int trans, char *line, s
 			}
 			line[strcspn (line, " \t=")] = '\0';
 			val = line + i + 1 + strspn (line + i + 1, " \t");
+			valsize = 0;
+			while (val[valsize++]);
 
-			/* Skip trailing comments, if any */
-			for (j = 0; j < sz - (val - line); j++) {
-				if (val[j] == '#' || val[j] == ';') {
+			/* Skip trailing spaces and comments */
+			for (j = 0; j < valsize; j++) {
+				if (val[j] == '#' || val[j] == ';' || isspace(val[j])) {
 					val[j] = '\0';
 					break;
 				}
 			}
-
-			/* Skip trailing whitespace, if any */
-			for (j--; j > 0; j--) {
-				if (isspace(val[j]))
-					val[j] = '\0';
-				else 
-					break;
-			}
-
 			/* XXX Perhaps should we not ignore errors?  */
 			conf_set(trans, section, arg, line, val, 0, 0);
 			return;
