diff -up nfs-utils-1.2.3/aclocal/libtirpc.m4.orig nfs-utils-1.2.3/aclocal/libtirpc.m4
--- nfs-utils-1.2.3/aclocal/libtirpc.m4.orig	2015-11-18 07:39:01.000000000 -0500
+++ nfs-utils-1.2.3/aclocal/libtirpc.m4	2015-11-18 08:59:26.887585298 -0500
@@ -53,6 +53,11 @@ AC_DEFUN([AC_LIBTIRPC], [
       AC_DEFINE([HAVE_AUTHGSS_FREE_PRIVATE_DATA], 1,
 	      [Define to 1 if your rpcsec library provides authgss_free_private_data,])
     fi
+	AS_IF([test -n "${LIBTIRPC}"],
+		[AC_CHECK_LIB([tirpc], [libtirpc_set_debug],
+			[AC_DEFINE([HAVE_LIBTIRPC_SET_DEBUG], [1],
+				[Define to 1 if your tirpc library provides libtirpc_set_debug])],,
+			[${LIBS}])])
   else
     LIBTIRPC=""
   fi
diff -up nfs-utils-1.2.3/support/include/nfslib.h.orig nfs-utils-1.2.3/support/include/nfslib.h
--- nfs-utils-1.2.3/support/include/nfslib.h.orig	2015-11-18 07:39:01.416688115 -0500
+++ nfs-utils-1.2.3/support/include/nfslib.h	2015-11-18 08:54:30.538318041 -0500
@@ -176,6 +176,9 @@ size_t  strlcpy(char *, const char *, si
 ssize_t atomicio(ssize_t (*f) (int, void*, size_t),
 		 int, void *, size_t);
 
+#ifdef HAVE_LIBTIRPC_SET_DEBUG
+void  libtirpc_set_debug(char *name, int level, int use_stderr);
+#endif
 
 #define UNUSED(x) UNUSED_ ## x __attribute__((unused))
 
diff -up nfs-utils-1.2.3/utils/gssd/gssd.c.orig nfs-utils-1.2.3/utils/gssd/gssd.c
--- nfs-utils-1.2.3/utils/gssd/gssd.c.orig	2015-11-18 07:39:01.374687149 -0500
+++ nfs-utils-1.2.3/utils/gssd/gssd.c	2015-11-18 08:54:30.543318146 -0500
@@ -178,6 +178,13 @@ main(int argc, char *argv[])
 	if (verbosity && rpc_verbosity == 0)
 		rpc_verbosity = verbosity;
 	authgss_set_debug_level(rpc_verbosity);
+#elif HAVE_LIBTIRPC_SET_DEBUG
+	/*
+	 * Only set the libtirpc debug level if explicitly requested via -r...
+	 * gssd is chatty enough as it is.
+	 */
+	if (rpc_verbosity > 0)
+		libtirpc_set_debug(progname, rpc_verbosity, fg);
 #else
         if (rpc_verbosity > 0)
 		printerr(0, "Warning: rpcsec_gss library does not "
diff -up nfs-utils-1.2.3/utils/gssd/svcgssd.c.orig nfs-utils-1.2.3/utils/gssd/svcgssd.c
--- nfs-utils-1.2.3/utils/gssd/svcgssd.c.orig	2015-11-18 07:39:01.375687172 -0500
+++ nfs-utils-1.2.3/utils/gssd/svcgssd.c	2015-11-18 08:54:30.543318146 -0500
@@ -221,6 +221,13 @@ main(int argc, char *argv[])
 	if (verbosity && rpc_verbosity == 0)
 		rpc_verbosity = verbosity;
 	authgss_set_debug_level(rpc_verbosity);
+#elif HAVE_LIBTIRPC_SET_DEBUG
+        /*
+	 * Only set the libtirpc debug level if explicitly requested via -r...
+	 * svcgssd is chatty enough as it is.
+	 */
+        if (rpc_verbosity > 0)
+                libtirpc_set_debug(progname, rpc_verbosity, fg);
 #else
 	if (rpc_verbosity > 0)
 		printerr(0, "Warning: rpcsec_gss library does not "
