diff -up nfs-utils-1.3.0/systemd/nfs-blkmap.service.orig nfs-utils-1.3.0/systemd/nfs-blkmap.service
--- nfs-utils-1.3.0/systemd/nfs-blkmap.service.orig	2014-03-25 11:12:07.000000000 -0400
+++ nfs-utils-1.3.0/systemd/nfs-blkmap.service	2014-12-09 13:25:56.145156424 -0500
@@ -13,4 +13,4 @@ PartOf=nfs-utils.service
 [Service]
 Type=forking
 PIDFile=/var/run/blkmapd.pid
-ExecStart=/usr/sbin/blkmapd
+ExecStart=/usr/sbin/blkmapd $BLKMAPDARGS
