diff -up nfs-utils-1.2.3/tools/mountstats/mountstats.py.orig nfs-utils-1.2.3/tools/mountstats/mountstats.py
--- nfs-utils-1.2.3/tools/mountstats/mountstats.py.orig	2015-03-04 08:16:24.361213487 -0500
+++ nfs-utils-1.2.3/tools/mountstats/mountstats.py	2015-03-04 08:16:45.449573239 -0500
@@ -676,7 +676,7 @@ def mountstats_command(args):
     """Mountstats command
     """
     mountstats = parse_stats_file(args.infile)
-    mountpoints = args.mountpoints
+    mountpoints = [os.path.normpath(mp) for mp in args.mountpoints]
 
     # make certain devices contains only NFS mount points
     if len(mountpoints) > 0:
@@ -725,7 +725,7 @@ def nfsstat_command(args):
     """nfsstat-like command for NFS mount points
     """
     mountstats = parse_stats_file(args.infile)
-    mountpoints = args.mountpoints
+    mountpoints = [os.path.normpath(mp) for mp in args.mountpoints]
     v3stats = DeviceData()
     v3stats.setup_accumulator(Nfsv3ops)
     v4stats = DeviceData()
@@ -812,7 +812,7 @@ def iostat_command(args):
     """iostat-like command for NFS mount points
     """
     mountstats = parse_stats_file(args.infile)
-    devices = args.mountpoints
+    devices = [os.path.normpath(mp) for mp in args.mountpoints]
 
     if args.since:
         old_mountstats = parse_stats_file(args.since)
